streamline stride propagation logic in TensorIterator

ngimel opened the pull request:
Fixes #41314 among other things. 
This PR streamlines layout propagation logic in TensorIterator and removes almost all cases of channels-last hardcoding. The new rules and changes are as follows:
1) behavior of undefined `output` and defined output of the wrong (e.g. 0) size is always the same (before this PR the behavior was divergent)
2) in obvious cases (unary operation on memory-dense tensors, binary operations on memory-dense tensors with the same layout) strides are propagated (before propagation was inconsistent) (see footnote)
3) in other cases the output permutation is obtained as inverse permutation of sorting inputs by strides. Sorting is done with comparator obeying the following rules: strides of broadcasted dimensions are set to 0, and 0 compares equal to anything. Strides of not-broadcasted dimensions (including dimensions of size `1`) participate in sorting. Precedence is given to the first input, in case of a tie in the first input, first the corresponding dimensions are considered, and if that does not indicate that swap is needed, strides of the same dimension in subsequent inputs are considered. See changes in `reorder_dimensions` and `compute_strides`. Note that first inspecting dimensions of the first input allows us to better recover it's permutation (and we select this behavior because it more reliably propagates channels-last strides) but in some rare cases could result in worse traversal order for the second tensor. 
 
These rules are enough to recover previously hard-coded behavior related to channels last, so all existing tests are passing.
In general, these rules will produce intuitive results, and in most cases permutation of the full size input (in case of broadcasted operation) will be recovered, or permutation of the first input (in case of same sized inputs) will be recovered, including cases with trivial (1) dimensions. As an example of the latter, the following tensor
```
x=torch.randn(2,1,3).permute(1,0,2)
```
will produce output with the same stride (3,3,1) in binary operations with 1d tensor. Another example is a tensor of size N1H1 that has strides `H,H,1,1` when contiguous and `H, 1, 1, 1` when channels-last. The output retains these strides in binary operations when another 1d tensor is broadcasted on this one.  

Footnote: for ambiguous cases where all inputs are memory dense and have the same physical layout that nevertheless can correspond to different permutations, such as e.g. NC11-sized physically contiguous tensors, regular contiguous tensor is returned, and thus permutation information of the input is lost (so for NC11 channels-last input had the strides `C, 1, C, C`, but output will have the strides `C, 1, 1, 1`). This behavior is unchanged from before and consistent with numpy, but it still makes sense to change it. The blocker for doing it currently is performance of `empty_strided`. Once we make it on par with `empty` we should be able to propagate layouts in these cases. For now, to not slow down common contiguous case, we default to contiguous. 
The table below shows how in some cases current behavior loses permutation/stride information, whereas new behavior propagates permutation. 
| code                                                                                                                                                                                           | old                                                   | new                                                  |
|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------|------------------------------------------------------|
| #strided tensors<br>a=torch.randn(2,3,8)[:,:,::2].permute(2,0,1)<br>print(a.stride())<br>print(a.exp().stride())<br>print((a+a).stride())<br>out = torch.empty(0)<br>torch.add(a,a,out=out)<br>print(out.stride()) | (2, 24, 8) <br>(6, 3, 1) <br>(1, 12, 4) <br>(6, 3, 1) | (2, 24, 8)<br>(1, 12, 4)<br>(1, 12, 4)<br>(1, 12, 4) |
| #memory dense tensors<br>a=torch.randn(3,1,1).as_strided((3,1,1), (1,3,3))<br>print(a.stride(), (a+torch.randn(1)).stride())<br>a=torch.randn(2,3,4).permute(2,0,1)<br>print(a.stride())<br>print(a.exp().stride())<br>print((a+a).stride())<br>out = torch.empty(0)<br>torch.add(a,a,out=out)<br>print(out.stride())                                                                                                                                                                                               |  (1, 3, 3) (1, 1, 1)<br>(1, 12, 4)<br>(6, 3, 1)<br>(1, 12, 4)<br>(6, 3, 1)                                                       |  (1, 3, 3) (1, 3, 3)<br>(1, 12, 4)<br>(1, 12, 4)<br>(1, 12, 4)<br>(1, 12, 4) |

Commit:
streamline stride propagation logic

dzhulgakov commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -302,53 +302,42 @@ DimVector TensorIterator::apply_perm_and_mul(IntArrayRef input, int mul) const {
+      }
+      auto tensor_shape = invert_perm(shape_);
+      if (inverted) {
+        if (!op.tensor.defined())
```
nit, add {}

dzhulgakov commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -302,53 +302,42 @@ DimVector TensorIterator::apply_perm_and_mul(IntArrayRef input, int mul) const {
+        // can just return contiguous output
+        // it is faster because it avoids allocating 0 size tensor and
+        // resizing and restriding it
+          op.tensor = at::empty(tensor_shape, op.options());
```
so we might just delete this branch if empty_strided is faster? maybe expand the comment

ngimel replied:
There's still unpermuting of strides that needs to be done even if empty_strided is faster, depending on perf impact we might ignore it and delete this branch.

dzhulgakov commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -302,53 +302,42 @@ DimVector TensorIterator::apply_perm_and_mul(IntArrayRef input, int mul) const {
+      op.stride_bytes = compatible_stride(element_size);
+      // check if permutation is just an inverted order
+      bool inverted = true;
+      for (int i = 1; i <= ndim(); i++) {
```
very nit: write loop from 0 and `<`, all loops starting from 1 are suspicious :)

dzhulgakov commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -791,6 +780,23 @@ void TensorIterator::mark_outputs() {
+        operands_[i].will_resize = true;
+        continue;
+      }
+      TORCH_CHECK(is_reduction_,  "output with shape ", output.sizes(), " doesn't match the broadcast shape ",
```
just curious - why is_reduction case is fine? maybe add a comment? (I know this logic predated this diff, but still)

dzhulgakov commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -47,9 +47,9 @@ void TensorIterator::reorder_dimensions(const TensorIteratorConfig& config) {
-      } else {
+      } else  if (stride0 > stride1) {
         return 1;
       }
```
so we treat equal dims as ambigious? shouldn't we do something like "preserver the original order", i.e. compare dim0 and dim1 themselves? Maybe that's what the insertion sort below does, but it's hard for me to mentally unravel it

ngimel replied:
This is a very good question! Added a "flies in the ointment" section to the description. When strides are equal, we don't swap corresponding dimension, so we tend to favor identity (or closer-to-identity) permutation in this case.

dzhulgakov reviewed (commented):
Epic! It would be a good idea to merge a lot of observations from the PR descriptions as inline comments in the right places. All this logic is way too magical and requires explanations :)

Commit:
fix sizes for intermediate tensors

Commit:
add test

Commit:
updates per review

Commit:
reinitialize out in test

VitalyFedyunin commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -893,64 +864,6 @@ void TensorIterator::compute_strides(const TensorIteratorConfig& config) {
   }
 }
 
-template <int dim, MemoryFormat memory_format>
```
Please remove next 3 functions from .h file as well.

VitalyFedyunin commented:
I'm impressed!

Couple questions:
- Have you tried to benchmark this implementation?
- Are we fine about dropping warnings in terms of contig + c_l operations?

ngimel commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -47,9 +47,9 @@ void TensorIterator::reorder_dimensions(const TensorIteratorConfig& config) {
       }
       if (stride0 == 0 || stride1 == 0) {
         continue;
-      } else if (stride0 <= stride1) {
```
With `<=` logic the original tests were not passing, the loop was breaking too early and full permutation was not recovered.

glaringlee replied:
I see why, thanks.

ngimel commented:
> * Have you tried to benchmark this implementation?
> * Are we fine about dropping warnings in terms of contig + c_l operations?

Fast path benchmarks https://www.internalfb.com/intern/aibench/details/2193418916/ - for some reason inplace ops regress ~5% which is very weird, given that code path is literally the same. I'll do benchmarks for slow paths.
I think it's ok to drop warnings, we just have generic rules for layout propagation now that is not specific for channels_last. Similarly, when there's an operation on tensors with different permutations, we don't warn.

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -842,35 +842,6 @@ void TensorIterator::compute_shape(const TensorIteratorConfig& config) {
-  if (config.static_shape_.has_value()) {
-    return;
-  }
-  // Outputs cannot be broadcasted. Check that the shape of the outputs matches
```
I noticed you didn't preserve this comment; intentional? I think it's still true that outputs cannot be broadcasted...

ngimel replied:
No, it's not intentional, I restore the comment.

ezyang commented:
At some point we should reorder the definitions in TensorIterators.cpp to be in order they run in the algorithm, it would have been a lot easier to read the PR if I could read top down and know that it was chronological

ezyang commented:
Note for which parts of TensorIterator algorithm got modified:

```
  // try fast setup output tensor, if failed, fallback to normal setup
  if (!fast_set_up(config)) { // MODIFIED
    // compute each tensor's stride after broadcasting
    compute_strides(config);  // MODIFIED
    // re-order dimensions to improve coalescing
    reorder_dimensions(config); // MODIFIED
    // allocate the output tensor if it's not provided
    allocate_outputs(); // RENAMED AND MODIFIED
    // coalesce adjacent dimensions when possible
    coalesce_dimensions();
  }

```

ezyang commented:
I was expecting to see something in the PR description talking about "resize", but I didn't see anything anywhere. Maybe the description can be expanded to talk about the general strategy that the new algorithm takes? (Or better yet, as Dmytro suggests, put it in code.)

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -1023,6 +936,8 @@ bool TensorIterator::fast_set_up(const TensorIteratorConfig& config) {
             TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", i);
             op.tensor = at::empty(shape_, op.options(), MemoryFormat::Contiguous);
             op.current_dtype = op.target_dtype;
+          } else if (op.will_resize) {
```
Do we have an invariant that it will never be the case that a tensor is undefined AND `will_resize` is true?

ngimel replied:
Yes, `will_resize` can be set to `true` only in `mark_resize_outputs`, and `tensor.defined()` is a necessary condition to enter that branch.

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -1053,9 +972,10 @@ bool TensorIterator::fast_set_up(const TensorIteratorConfig& config) {
             op.current_dtype = op.target_dtype;
+          } else if (op.will_resize) {
+            at::native::resize_output(op.tensor, shape_);
+            op.tensor.as_strided_(shape_, operands_[i_defined].tensor.strides());
```
Trusting that these are the only three sites that need updating.

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -883,7 +854,7 @@ void TensorIterator::compute_strides(const TensorIteratorConfig& config) {
           op.stride_bytes.resize(ndim());
       for (size_t i = 0; i < original_shape.size(); i++) {
-        if (original_shape[i] == 1) {
+        if (original_shape[i] == 1 && shape_[offset + i] !=1) {
```
This would be a great place to have a reference to a Note explaining the overall algorithm

ezyang replied:
"strides of broadcasted dimensions are set to 0, and 0 compares equal to anything. Strides of not-broadcasted dimensions (including dimensions of size 1) participate in sorting."

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -645,6 +625,9 @@ void TensorIterator::cast_outputs() {
         op.original_tensor.scalar_type() != op.current_dtype) {
+      if (op.original_tensor.sizes() != op.tensor.sizes()){
+        op.original_tensor.resize_as_(op.tensor).as_strided_(op.tensor.sizes(), op.tensor.strides());
+      }
```
Is this because we are lazily resizing now? I'm surprised that `allocate_or_resize_outputs` cannot be assumed to have handled this for you.

ngimel replied:
Yes, we used to resize before computing types, so original output was resized to correct size (but not necessarily correct layout) before intermediate output was allocated. Now, by the time we are resizing, original output is forgotten and we resize intermediate only. 
That's a good point though, I better handle this in `allocate_or_resize_outputs` anyway.

ezyang commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -293,62 +293,42 @@ DimVector TensorIterator::invert_perm(IntArrayRef input) const {
-        // make sure that operand's strides are matching element size and
-        // dimensions permutations which are stored in _perm
-        op.stride_bytes = apply_perm_and_mul(op.tensor.strides(), element_size);
-      } else {
```
So sweet, so savory.

ezyang commented on test/test_torch.py:
```diff
@@ -15008,6 +15008,62 @@ def _test_helper(x, y, bias, memory_format):
             torch.randn((1, 3, 1, 1, 1), device=device).contiguous(memory_format=torch.channels_last_3d),
             torch.channels_last_3d)
 
+    def test_strides_propagation(self, device):
```
cc @mruberry on this test

ezyang commented on test/test_torch.py:
```diff
@@ -15008,6 +15008,62 @@ def _test_helper(x, y, bias, memory_format):
+
+            dim = x.dim()
+            div = x.stride(dim - 1)
+            for p in permutations(range(dim)):
```
Nice!

ezyang commented on test/test_torch.py:
```diff
@@ -15008,6 +15008,62 @@ def _test_helper(x, y, bias, memory_format):
+        binary_ops = (torch.eq, torch.add)
+        unary_ops = (torch.exp,)
+        # memory dense and sliced
+        xs = (torch.randn(2, 3, 4, device=device), torch.randn(2, 3, 8, device=device)[:, :, ::2])
```
No testing for the ambiguous cases?

ngimel replied:
Added a test for strided ambiguous case

ezyang reviewed (approved):
This is all super subtle, but it seems to make sense, and magically the code gets smallers. Nice work! Looking for more comments, but if you want to perhaps defer that for a second PR that's fine by me too.

Commit:
Merge branch 'master' into ti_prop

Commit:
adds note, inline comments, addresses other review comments

Commit:
Merge branch 'master' into ti_prop

Commit:
lint

ngimel commented:
Slightly changed the algorithm to better propagate input strides in ambiguous cases, changed the description to reflect that, added a note in the code, added a test with ambiguous sizes.
